homeworks/solutions/krasnor/5-remesh/src/halfedge.js

```diff
@@ -0,0 +1,821 @@
+      return value.idx >= 0;
+    });
+
+    // reset indices
+    for(let i = 0; i < this.edges.length; i++){
+      this.edges[i].idx = i;
+    }
+    for(let i = 0; i < this.halfedges.length; i++){
+      this.halfedges[i].idx = i;
+    }
+    for(let i = 0; i < this.vertices.length; i++){
+      this.vertices[i].idx = i;
+    }
+    for(let i = 0; i < this.faces.length; i++){
+      this.faces[i].idx = i;
+    }
+    console.log("Vertices: %d, Edges: %d, Faces: %d", this.vertices.length, this.edges.length, this.faces.length);
+
+    // end of simplify
+  }
+
+  // /**
+  //  *
+  //  * @param {Matrix} Q
+  //  * @param {Vector} v
+  //  */
+  // calc_error(Q, v){
+  //   let err =
+  //       Q.x00 * v.x*v.x  + 2*Q.x01 * v.x*v.y  + 2*Q.x02 * v.x*v.z  + 2*Q.x03 * v.x +
+  //                            Q.x11 * v.y*v.y  + 2*Q.x12 * v.y*v.z  + 2*Q.x13 * v.y +
+  //                                                 Q.x22 * v.z*v.z  + 2*Q.x23 * v.z +
+  //                                                                      Q.x33;  // = v^t*Q*v
+  //   return err;
+  // }
+  /**
+   *
+   * @param {Matrix} Q
+   * @param {Vector} v point
+   */
+  calc_error_v2(Q, v){
+    // v^t * Q * v = v^t * right
+
+    // Q * v
+    let right = Q.mulpos(v);
+    let w = Q.x30*v.x + Q.x31*v.y + Q.x32*v.z + Q.x33;
+
+    // v^t * right
+    let err = v.x * right.x + v.y * right.y + v.z * right.z + w;
+    return err;
+
+
+    //                  (x)
+    //                  (y)
+    //                  (z)
+    // (x,y,z,w) * Q *  (w) = v^t * Q * v = v^t * r
+
+    //              (r.x)
+    //              (r.y)
+    //              (r.z)
+    // (x,y,z,w) *  (r.w) = v^t * r = v^t * (Q*v)
+  }
+
+  /**
+   *
+   * @param q1 Matrix
+   * @param q2 Matrix
+   * @returns {{mergePoint: Vector, error: Number}}
+   */
+  computeEdgeError(q_1,q_2, edge){
+    let Qbar = q_1.add(q_2);
+    let dbg_Qbar_Invertible = false;
+    // compute vbar
+    let vbar = new Vector();
+
+    // if(Qbar.det() != 0){
+    if (Math.abs(Qbar.det()) > 1e-3) {
+      dbg_Qbar_Invertible = true;
+      // invertible, one solution
+      Qbar.x30 = 0;
+      Qbar.x31 = 0;
+      Qbar.x32 = 0;
+      Qbar.x33 = 1;
+      vbar = Qbar.inv().mulpos(new Vector(0,0,0));
+    }else{
+      // not invertible, 0 or infinite solutions => approximate
+      dbg_Qbar_Invertible  = false;
+
+      const n = 16
+      const a = edge.halfedge.vertex.position
+      const b = edge.halfedge.next.vertex.position
+      const d = b.sub(a)
+      let beste = -1.0
+      let bestv = new Vector()
+      for (let i = 0; i <= n; i++) {
+        const t = i / n
+        const v = a.add(d.scale(t))
+        const e = this.calc_error_v2(Qbar,v) // you need implement this yourself, which is calculating v^T q v :)
+        if (beste < 0 || e < beste) {
+          beste = e
+          bestv = v
+        }
+      }
+      vbar = bestv
+
+      // alternative - take mid point, or one of the edge points
```

Comment:
How's the result looks like? Does searching on mid-point more efficient and resulting in similar results?